src/db.c

```diff
@@ -980,13 +981,24 @@ void randomkeyCommand(client *c) {
 void keysCommand(client *c) {
     dictEntry *de;
     sds pattern = c->argv[1]->ptr;
-    int plen = sdslen(pattern), allkeys;
+    int plen = sdslen(pattern), allkeys, pslot = -1;
     long numkeys = 0;
     void *replylen = addReplyDeferredLen(c);
-    dbIterator *dbit = dbIteratorInit(c->db, DB_MAIN);
     allkeys = (pattern[0] == '*' && plen == 1);
+    dbIterator *dbit;
+    if (server.cluster_enabled) {
+        if (!allkeys) {
+            pslot = patternHashSlot(pattern, plen);
+        }
+        dbit = dbIteratorInitFromSlot(c->db, DB_MAIN, pslot > -1 ? pslot - 1 : -1);
```

Comment:
Init from `pslot - 1` seems wrong. Init from `pslot`?

Also checking for valid slot I think >= 0 looks more intuitive than > -1.

```suggestion
        dbit = dbIteratorInitFromSlot(c->db, DB_MAIN, pslot >= 0 ? pslot : -1);
```

Reply:
If we init form `pslot`, the iterarition will start from slot `pslot+1`, but we need to iterate the keys in slot `pslot`.

Reply:
Really? I think it's an unintuitive, strange API. Maybe we can change it? or maybe we can add at least some comment about it above the definition dbIteratorInitFromSlot or dbIteratorNext?

Reply:
@hpatro please take a look